File: java/StrataServer/src/subsystems/strata1/server/bootstrap/ServerBootstrapper.java
// ##########################################################################
// # File Name:	ServerBootstrapper.java
// #
// # Copyright:	2013, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataServer Framework.
// #
// #   			The StrataServer Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataServer Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataServer
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.server.bootstrap;

import java.util.Collections;
import java.util.List;
import strata1.common.logger.ILogger;
import strata1.common.logger.LoggingLevel;
import strata1.common.producerconsumer.ITaskProducerConsumerManager;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class ServerBootstrapper
    implements IServerBootstrapper
{
    private IServerContainer             itsContainer;
    private ILogger                      itsLogger;
    private ITaskProducerConsumerManager itsTaskManager;
    private List<IServerModule>          itsModules;
    private boolean                      itsStartIndicator;
    
    /************************************************************************
     * Creates a new {@code ServerBootstrapper}. 
     *
     */
    public
    ServerBootstrapper()
    {
        this( false );
    }
    
    /************************************************************************
     * Creates a new {@code ServerBootstrapper}. 
     *
     * @param start
     */
    public 
    ServerBootstrapper(boolean start)
    {
        itsContainer      = null;
        itsLogger         = null;
        itsTaskManager    = null;
        itsModules        = null;
        itsStartIndicator = start;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public IServerContainer 
    getContainer()
    {
        return itsContainer;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public ILogger 
    getLogger()
    {
        return itsLogger;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public ITaskProducerConsumerManager 
    getTaskManager()
    {
        return itsTaskManager;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public List<IServerModule> 
    getModules()
    {
        return Collections.unmodifiableList( itsModules );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    run(IServerFactory factory)
    {
        try
        {
            CreateContainer(factory);
            CreateLogger(factory);
            CreateTaskManager(factory);
            CreateModules(factory);
            InitializeModules();
            
            if ( MustStartServer() )
                StartServer();
        }
        catch (Exception e)
        {
            if ( getLogger() != null )
                getLogger().logError( e.getMessage() );
        }
    }

    /************************************************************************
     *  
     *
     * @param factory
     */
    private void 
    CreateContainer(IServerFactory factory)
    {
        itsContainer = factory.createServerContainer();
        
        if ( getContainer() == null )
            throw new IllegalStateException("container is null");
    }

    /************************************************************************
     *  
     *
     * @param factory
     */
    private void 
    CreateLogger(IServerFactory factory)
    {
        itsLogger = factory.createLogger();
        
        if ( getLogger() == null )
            throw new IllegalStateException("logger is null");
        
        getLogger().logInfo( "Registering logger with container" );
        //getContainer().registerLogger(itsLogger);
    }

    /************************************************************************
     *  
     *
     * @param factory
     */
    private void 
    CreateTaskManager(IServerFactory factory)
    {
        getLogger().logInfo( "Creating task manager." );
        itsTaskManager = factory.createTaskProducerConsumerManager();
        
        if ( getTaskManager() == null )
            throw new IllegalStateException( "task manager is null" );
        
        getLogger().logInfo( "Creating task consumer." );
        getTaskManager().attachConsumer( factory.createTaskConsumer() );
        
        getLogger().logInfo( "Registering task manager with container." );
        //getContainer().registerTaskProducerConsumerManager(getTaskManager());
    }

    /************************************************************************
     *  
     *
     * @param factory
     */
    private void 
    CreateModules(IServerFactory factory)
    {
        getLogger().logInfo( "Creating server modules." );
        itsModules = factory.createModules();
        
        if ( getModules() == null )
            throw new IllegalStateException( "modules is null." );
        
        if ( getModules().isEmpty() )
            throw new IllegalStateException( "no modules were created." );
    }

    /************************************************************************
     *  
     *
     */
    private void 
    InitializeModules()
    {
        getLogger().logInfo( "Initializing server modules." );
        
        for (IServerModule module : getModules())
        {
            getLogger()
                .logInfo("Initalizaing module: " + module.getName() + ".");
            module.initialize( this );
        }
    }

    /************************************************************************
     *  
     *
     * @return
     */
    private boolean 
    MustStartServer()
    {
        return itsStartIndicator;
    }

    /************************************************************************
     *  
     *
     */
    private void 
    StartServer()
    {
        getTaskManager().startUp();
    }


}

// ##########################################################################

File: legacy/csharp/components/StrataServer/Bootstrapper/ServerBootstrapper.cs
//  ##########################################################################
//  # File Name: AbstractServerBootstrapper.cs
//  # Copyright: 2012, Capital Group Companies, Inc.
//  ##########################################################################

using System;
using System.Collections.Generic;
using Strata.Common.Logging;
using Strata.Common.Task;
using Microsoft.Practices.Unity;

namespace Strata.Server.Bootstrapper
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public
    class ServerBootstrapper:
        IServerBootstrapper
    {
        public ILogger                    Logger { get; protected set; }
        public IUnityContainer            Container { get; protected set; }
        public IList<IServerModule>       Modules { get; protected set; }
        public IStartupShutdownController StartupShutdownController{get;protected set;}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public 
        ServerBootstrapper() {}

        //////////////////////////////////////////////////////////////////////
        /// <see cref="IServerBootstrapper.Run(IServerFactory)"/>
        /// 
        public virtual void 
        Run(IServerFactory factory)
        {
            try
            {
                CreateContainer(factory);
                CreateLogger(factory);
                CreateModules(factory);
                CreateStartupShutdownController(factory);
                InitializeModules();
                ConfigureStartupShutdownController();
            }
            catch (Exception e)
            {
                if ( Logger != null )
                    Logger.Error(e.Message);

                throw e;
            }   
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        CreateContainer(IServerFactory factory)
        {
            Container = factory.CreateContainer();

            if ( Container == null )
                throw new NullReferenceException("container is null");
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        CreateLogger(IServerFactory factory)
        {
            Logger = factory.CreateLogger();

            if ( Logger == null )
                throw new NullReferenceException("logger is null");

            Logger.Verbose("Registering logger with container.");
            Container.RegisterInstance<ILogger>(Logger);
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        CreateModules(IServerFactory factory)
        {
            Logger.Verbose("Creating server modules.");
            Modules = factory.CreateModules();

            if ( Modules == null )
                throw new NullReferenceException("modules is null");

            if ( Modules.Count == 0 )
                throw new InvalidOperationException("no modules were created");
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        CreateStartupShutdownController(IServerFactory factory)
        {
            Logger.Verbose("Creating server startup controller.");
            StartupShutdownController = 
                factory.CreateStartupShutdownController();

            if ( StartupShutdownController == null )
                throw 
                    new NullReferenceException(
                        "startup-shutdown controller is null");
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        InitializeModules()
        {
            Logger.Verbose(
                "\n" +                
                "*******************************************************\n" + 
                "************* Initializing server modules *************\n" +
                "*******************************************************\n");

            foreach (IServerModule module in Modules)
            {
                Logger.Verbose("Initalizing module: " + module.Name + ".");
                module.Initialize(this);
            }   
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        protected virtual void 
        ConfigureStartupShutdownController()
        {
            StartupShutdownController.Container = Container;
        }
    }
}

//  ##########################################################################
